### DISCUSSION TITLE
关于如何实现 QA 调整服务器时间的需求

### DISCUSSION THREAD

**Discussion by tqing1128** [cloudwu/skynet#1869] [General]
我使用 docker 来部署测试环境，一台宿主机，N 个环境，每个 QA 一个环境（一组容器），需要提供 QA 改自己环境时间但是不影响其他环境（一组容器）的功能。

我使用 [libfaketime](https://github.com/wolfcw/libfaketime) 是可以实现该功能的，但是当我使用 jemalloc 后，就出现冲突了，原因如下：

[issues](https://github.com/wolfcw/libfaketime/issues/130)

还有其他方式实现该需求吗

**Comment by cloudwu** [comment]
skynet 整个系统都不依赖 walltime ，你关于时间的需求应该是自己调用系统函数实现的。

你不需要修改系统时间，而应该自己实现一个管理时间任务的服务，类似 crontab 。给这个服务增加调试功能。

**Reply by tqing1128** [reply]
crontab 只能触发，只能解决部分，有些逻辑中使用的是时间戳，比如 `每日重置` 类业务，通常是记录上一次刷新的时间戳，然后与当前时间戳对比，是不同的 `天` 才刷新，`天` 发生变化时触发这个逻辑，如果时间没有真正的变化，业务是不会重置的

**Reply by huahua132** [reply]
https://github.com/huahua132/skynet_fly
可以考虑用我的这种加速时间方式

**Reply by tqing1128** [reply]
这种方式会不会影响容器宿主机的时间？

**Reply by huahua132** [reply]
这种方式是通过加速skynet内部定时器时间滴答数，不会影响到系统，像os.time不会受影响

**Reply by tqing1128** [reply]
我看到会执行 http://127.0.0.1:debug_port/fasttime/... ，最终是怎么实现的

**Comment by huahua132** [comment]
需要改动到源码，具体改动可以查看`https://github.com/cloudwu/skynet/pull/1812`。改动好的fork版本：`https://github.com/huahua132/skynet`

**Reply by tqing1128** [reply]
明白了，谢谢你的方案

**Comment by tqing1128** [comment]
上面扩展 skynet 是一个解决方案，但是正如云风大哥所说，这会带来额外的维护成本，对于 skynet 框架的使用者也是一样的。我更偏向于使用不修改 skynet 源码的方案，就像是 libfaketime，有没有类似的又不和 jemalloc 冲突的方案？

**Reply by ylme** [reply]
自己维护这个 offset_sec ，可以试试这种方式。

```lua
function skynet.current_time()
    return skynet.time() // 1 + offset_sec
end
```

**Reply by tqing1128** [reply]
这需要广播到所有服务，而且 skynet.timeout 不会触发

**Reply by fanyh** [reply]
重写skynet.time 是比较理想的解决方案
如果期望timeout 也可以命中，那就只有魔改skynet_updatetime；
btw:事实上所有的业务层应当如云大侠所诉，你应该自己实现一个date, 在这个date上再实现调试需求就很容易了

time.lua
```
local T = {}
local offset = Skynet.getenv("offset") // 也可以用sharetable 实现
local starttime = Skynet.starttime()
function T.now()
    return Skynet.now()/100 + starttime + offset
end

function T.offset(time)
    local diff= time - T.now()
    assert(diff >= 0, "调试时间逆转 容易出现诸多奇怪时序问题，不建议逆转")
    offset = diff
end

return T
```

**Reply by tqing1128** [reply]
我前面有说明了，并非所有的业务都适合用 crontab 来实现的，被动地使用时间戳来刷新业务是非常常见的实现。crontab + offset 是一个方案，但是需要广播来同步，无法做到某一个时刻所有服务时间同时变化，需要先更新业务服务，再更新 crontab 服务，即使是这样都还可能会出现 bug。

**Reply by cloudwu** [reply]
为什么不适合用 crontab 实现？这里说的 crontab 不是一个独立进程，而是类似独立 crontab 的一个 skynet 内置服务。

无论你怎么实现业务，无非都是有若干地方会去获取和时间有关的信息，然后做对应的事情。如果你任何地方都不从系统调用取时间，那么自然业务就和系统时间无关。skynet 本身没有分散在不同地方的和系统时间有关的代码，所以做这个事情和 skynet 自己的实现没什么关系。

如果你的系统设计是良好的，那么，所有 `时刻 -> 业务` 都应该用 skynet 消息驱动，即到点了，产生了一个消息，消息让对应的业务发生。而管理这些消息的就应该在一个独立的 skynet 服务。你让这个服务支持测试环境，它就应该可以满足你调整时间的需求。

**Reply by tqing1128** [reply]
你说的我理解，我用适合来描述不恰当，应该说是更优先选择 `被动式刷新` 来实现，所谓 `被动式刷新` 简单来说就是需要使用数据的时候才会刷新。

拿比较典型的 `体力回复类` 业务来说，只要有 `初始回复时间` `回复间隔时间` `每次回复点数` 以及 `上限`，再加上服务器时间，客户端就可以做到自己计算，这是准确的，而真正使用数据的频率是比回复频率要小得多的，服务器就可以节省每次回复时的消耗，也不需要同步数据给客户端。

虽然节省的单个消耗是极小的，但是这是会叠加的，游戏业务会在 0 点时候达到高峰，`按小时刷新` `按天刷新` `按周刷新` `按月刷新` `活动开启` `活动结算` 等业务集中触发，在线人数越多，消耗也越多，内存和 cpu 都暴涨，`被动式刷新` 可以减少部分消耗，这是选择它的主要原因。

另外，对于 agent 会释放的业务框架，存在需要刷新较长时间段，可能出现体力从 0 回满，这种情况使用 crontab 服务实现会在很短的时间触发多次。

**Reply by cloudwu** [reply]
你说的这些和上面说的实现方法没有什么冲突。

你要的是 **服务器时间** 并且你希望它可以被测试环境调整。那么，就应该增加调整 **时间** 的方法。

**服务器时间** 在这里是一个参数，它会参与你的业务逻辑。任何参数，都不应该用共享状态的方式工作。即，**不应该设计一个全局唯一的状态** ，一经修改，所有地方都可以共享这个修改过程。

如果你需要在每个服务中都有一个 **服务器时间** 这个状态量，那么，就应该让这个状态在每个服务中都是独立自有的。这个状态量其实是你的业务逻辑的调整 + skynet.now() 获取的服务器开启时长的结果。前面这个调整值应该是由 crontab 服务推送给每个需要它的服务。如果没有调整过，那么就不需要推送。

**Reply by cloudwu** [reply]
**时间** 和别的什么参数相比，不应该有什么特殊性。就好比你有一个叫做 **攻击力修正系数** 的东西，不应该设计成一个全局变量，所有服务都可以读到这个变量，而允许你在一个地方调整，其它地方都自动能感知到。而应该用消息推送这个修改，每个服务独立有一份 **攻击力修正系数** 。

**服务器时间** 和 **攻击力修正系数** 对你的游戏业务来说，是同一类东西。时间没有特殊性。不要混淆游戏业务用到的时间和服务器底层的时间。

**Reply by tqing1128** [reply]
你说的混淆，按的我理解是业务未在事务过程中保持 **可变全局参数** 的一致性，为了保持一致性应该在事务开始或者第一次使用时拷贝，和参数是服务内的全局变量还是进程的全局变量没有关系。

要实现我的需求，需要串行执行 2 步：
1. 同步 **时间参数** 到所有服务，包括 crontab
2. 刷新 crontab

可通过修改服务器底层的时间以及 skynet 自动校时来实现，这只是一种实现方式。

由 crontab 广播推送的实现方式，对于夸服务和进程的事务，广播虽不能保证不同服务的参数一致性，但是事务自己保证好应该也是没有问题的。

但是我第一感觉是有问题的，我思考为什么会有这种感觉，感觉来源经验，应该是我经历的项目中忽略 **时间参数** 是一个  **可变全局参数** 情况很常见，不知道这是否是行业普遍存在。实际的项目中不一定有时间、人力去重构，项目也不一定能接受重构的风险，从这个角度看，修改服务器底层的时间的方案容错率更高点。

对于除了 skynet 框架，还使用其他框架的项目来说，修改服务器底层的时间的方案也通用些。

**Reply by cloudwu** [reply]
在 skynet 模式下，应该尽可能的避免全局状态。所有状态同步都应该通过消息。**时间** 没有特殊性。

如果你非要用一个全局 api 取业务用的时刻，那和 skynet 没什么关系。skynet 本身的运作依赖的是系统时间的单调递增，底层关心的是时间的间隔。没有任何地方关心绝对时刻。

自己封装一个取时刻的函数，自己管理好就可以了。

**Reply by sniper00** [reply]
我的方案是提供一个全局取时间函数, 很方便满足 测试人员测试时间相关的活动, 提供命令修改 offset_ 即可
```cpp
    static std::time_t now()
    {
           auto diff = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now() - start_time_point_);
           return start_millsecond + diff.count() + offset_;
    }
```

**Reply by fanyh** [reply]
我大概理解题主的意思是：业务中直接使用了 skynet自带的时间，基于各种原因又不打算修改实现方式
那么不外乎两种解决方案

1.  重写skynet引用,只需修改业务层的require
```
local skynet = require "skynet"

local start = skynet.starttime()

local myskynet = {}
function mysknet.time()
   return skynet.now() + start + offset
end
......
return myskynet
```

2. 魔改skynet_timer, timer 已经超10+年没有发生过变更，代码也足够简单，也无需顾忌冲突问题
```
void skynet_updatetime(void) {
    uint64_t cp = gettime();
    const char * offset = skynet_getenv("offset")
    cp += int(offset)
    ......
}
```

ps: 以上伪代码仅作为参考思路，未深入考虑严谨性

**Reply by huahua132** [reply]
在`skynet_updatetime`直接加 offset的方式没法让短间隔的定时器按照逻辑预期连续触达
```lua
while true do
    skynet.sleep(6000)
end
```
比如以上代码，如果直接设置`offset`为1小时，跨越的时间内，函数预期触达60次，这种修改方式只能触达一次

**Reply by tqing1128** [reply]
你这种情况只触发一次才合理吧

**Reply by tqing1128** [reply]
我漏了一点需求，调整时间后需要持久化，进程重启甚至是主机重启也要保持。

**Reply by huahua132** [reply]
正常流逝是触发60次，加速时间后只触发一次还合理吗

**Reply by fanyh** [reply]
持久化那就用上redis 缓存offset憋，讲真你要真这么干，还不如直接修改系统时间算了

**Reply by tqing1128** [reply]
是想修改系统时间，但是在 docker 容器内修改系统时间会影响宿主机的系统时间

**Reply by tqing1128** [reply]
> 正常流逝是触发60次，加速时间后只触发一次还合理吗

真实的业务中，应该是 crontab 触发，带上时间参数，业务自己结算，就好比体力回复，通过时间参数计算需要加 60 点，而不是执行 60 次加 1 点体力

**Reply by huahua132** [reply]
触发60次，又不会导致你体力多加。时间可以1秒1秒的快进。
不是所有需求都可以通过时间差计算。比如状态机类的游戏：
定时器A触发后，有可能注册定时器B或者定时器C。假设定时器A需要30分钟后触发，BC也是30分钟。用以上方式加速1小时只会触发A，而B或者C还需要等30分钟才能触发。

**Reply by fanyh** [reply]
可以试试用 [faketime](https://github.com/wolfcw/libfaketime)来修改docker中的时间，不影响宿主机
btw：对docker中的进程产生的中间值（offset）做持久化，这不奇怪么？按题主思维来：offset是对docker的时间进行修饰，只不过刚好表现在skynet上

**Reply by tqing1128** [reply]
你看看我最初的问题

**Reply by tqing1128** [reply]
> 触发60次，又不会导致你体力多加。时间可以1秒1秒的快进。 不是所有需求都可以通过时间差计算。比如状态机类的游戏： 定时器A触发后，有可能注册定时器B或者定时器C。假设定时器A需要30分钟后触发，BC也是30分钟。用以上方式加速1小时只会触发A，而B或者C还需要等30分钟才能触发。

你应该看看前面关于时间参数的讨论

**Reply by huahua132** [reply]
我看过了，你是想要一个进程内全局，可加速到某个时间节点，不影响系统时间的时间参数`skynet.time()`
然后还想要同时刷新crontab(我可以理解为触发skynet定时器事件不？)
我修改skynet底层的方案就可以做到

之间我的例子说明，是基于仅仅修改`skynet_updatetime()`后，去调整时间会存在的问题，因为一开始我的想法就是这样做的

**Comment by 253980289** [comment]
我的方案是自己重新封装了一套时间系统，也就是实际上有两套时间机制，一套是稳定的固定时间，另一套设计一个时间偏移值(默认为0)，支持gm修改这个偏移值供QA测试，所有应用逻辑使用这套偏差时间，框架系统级使用稳定时间(比如日志和db记录等)；同时还包装了一套定时器和周期任务系统(都是lua层模块)，当gm修改时间时，均会重新计算这些还未执行到的定时器重新确定执行时间(比如gm修改为快进了1天，那么每天执行一次的任务将会被立即执行一次，然后再计算下一天的执行时间)。
当然，这套机制并不完美，比如上面的例子中，对于每天计划0点执行一次的任务，如果gm直接修改快进5天，那么我们的任务系统只会执行一次，会少执行了4次，当然，如果真正需要到这个级别，你也可以将自定义的定时任务系统改为按时间差来执行跳过的所有执行次数。

**Reply by 397786756** [reply]
那这套方案的实现有demo吗?
